File: javascript/src/services/variables.ts
/* eslint-disable require-await -- Every method here validates its arguments before
   reaching the transport. Keeping them `async` makes a bad argument reject the
   returned promise rather than throw synchronously, so one `.catch()` covers both
   client-side and server-side failures. */

/** Variables service namespace — non-sensitive project configuration, held in plaintext. */

import { resolveProjectId, type ResolvedConfig } from "../config";
import type { Transport } from "../transport";
import {
  requireEntryName,
  requireEntryValue,
  type Actor,
  type ProjectEnvironment,
} from "./project-config";

/**
 * A project variable: non-sensitive configuration held in plaintext.
 *
 * Unlike a secret, the value is returned in full on every read. That is the
 * difference between the two resources, not an oversight — anything worth
 * hiding belongs in `client.secrets`.
 */
export interface ProjectVariable {
  id: string;
  /** Uppercase, underscore-separated, e.g. `LOG_LEVEL`. */
  name: string;
  description: string | null;
  value: string;
  environment: ProjectEnvironment;
  /** Id of the member who created it. */
  createdBy: string;
  /** Id of the member who last changed it. */
  updatedBy: string;
  createdByUser: Actor;
  updatedByUser: Actor;
  createdAt: string;
  updatedAt: string;
}

/** Options accepted by every method, for working across projects. */
export interface ProjectScopedParams {
  /** Overrides the client-level default project for this call. */
  projectId?: string;
}

/** Arguments accepted by `client.variables.list()`. */
export interface ListVariablesParams extends ProjectScopedParams {
  /** Limits the list to one environment. Omit to list them all. */
  environment?: ProjectEnvironment;
}

/** Arguments accepted by `client.variables.create()`. */
export interface CreateVariableParams extends ProjectScopedParams {
  /** Uppercase, underscore-separated, e.g. `LOG_LEVEL`. */
  name: string;
  value: string;
  description?: string;
  /** Defaults to `default` server-side. */
  environment?: ProjectEnvironment;
}

/**
 * Arguments accepted by `client.variables.update()`.
 *
 * Both fields are optional, but at least one has to be present — unlike a
 * secret, a variable can have its description changed without resetting its
 * value.
 */
export interface UpdateVariableParams extends ProjectScopedParams {
  value?: string;
  description?: string;
}

interface Envelope<T> {
  success: true;
  data: T;
}

export type ListVariablesResponse = Envelope<ProjectVariable[]>;
export type GetVariableResponse = Envelope<ProjectVariable>;
export type SetVariableResponse = Envelope<ProjectVariable>;

/**
 * Manages a project's variables.
 *
 * Reads need `variables:read`. Writes need `variables:write` *and* an owner or
 * admin: the API checks the member's role independently of the key's scopes,
 * so a plain member's key is refused even when it carries the scope.
 */
export class VariablesService {
  constructor(
    private readonly transport: Transport,
    private readonly config: ResolvedConfig
  ) {}

  /** Builds the collection route for the resolved project. */
  private basePath(projectId?: string): string {
    return `/v1/projects/${resolveProjectId(this.config, projectId)}/variables`;
  }

  /** Lists the project's variables, values included. */
  async list(params: ListVariablesParams = {}): Promise<ListVariablesResponse> {
    return this.transport.request<ListVariablesResponse>("GET", this.basePath(params.projectId), {
      query: { environment: params.environment },
    });
  }

  /** Fetches one variable, value included. */
  async get(variableId: string, params: ProjectScopedParams = {}): Promise<GetVariableResponse> {
    if (!variableId) throw new Error("variableId is required");

    return this.transport.request<GetVariableResponse>(
      "GET",
      `${this.basePath(params.projectId)}/${variableId}`
    );
  }

  /** Stores a new variable. */
  async create(params: CreateVariableParams): Promise<SetVariableResponse> {
    requireEntryName(params.name);
    requireEntryValue(params.value);

    return this.transport.request<SetVariableResponse>("POST", this.basePath(params.projectId), {
      body: {
        name: params.name,
        value: params.value,
        description: params.description,
        environment: params.environment,
      },
    });
  }

  /** Changes a variable's value, its description, or both. */
  async update(variableId: string, params: UpdateVariableParams): Promise<SetVariableResponse> {
    if (!variableId) throw new Error("variableId is required");
    if (params.value === undefined && params.description === undefined) {
      throw new Error("Provide a value or description to update");
    }
    if (params.value !== undefined) requireEntryValue(params.value);

    return this.transport.request<SetVariableResponse>(
      "PATCH",
      `${this.basePath(params.projectId)}/${variableId}`,
      { body: { value: params.value, description: params.description } }
    );
  }

  /** Permanently removes a variable. The API answers 204, so there is nothing to return. */
  async delete(variableId: string, params: ProjectScopedParams = {}): Promise<void> {
    if (!variableId) throw new Error("variableId is required");

    await this.transport.request<void>(
      "DELETE",
      `${this.basePath(params.projectId)}/${variableId}`
    );
  }
}

File: python/src/cosmoner/variables.py
"""Variables service namespace — non-sensitive configuration, held in plaintext."""

from __future__ import annotations

import re
from typing import Any

from ._config import ClientConfig, resolve_project_id
from ._transport import AsyncTransport, Transport

_NAME_PATTERN = re.compile(r"^[A-Z][A-Z0-9_]*$")
_MAX_NAME_LENGTH = 100
_MAX_VALUE_LENGTH = 10_000


def _require_variable_id(variable_id: str) -> None:
    """Rejects an empty variable id before it becomes a malformed route."""
    if not variable_id:
        raise ValueError("variable_id is required")


def _require_name(name: str) -> None:
    """Rejects a name the API would reject, using the message the API would send."""
    if not name:
        raise ValueError("name is required")
    if not _NAME_PATTERN.match(name):
        raise ValueError(
            "Name must be uppercase letters, numbers, or underscores, "
            "and start with a letter (e.g. LOG_LEVEL)"
        )
    if len(name) > _MAX_NAME_LENGTH:
        raise ValueError(f"name must be at most {_MAX_NAME_LENGTH} characters")


def _require_value(value: str) -> None:
    """Rejects a value the API would reject."""
    if not value:
        raise ValueError("value is required")
    if len(value) > _MAX_VALUE_LENGTH:
        raise ValueError(f"value must be at most {_MAX_VALUE_LENGTH} characters")


def _environment_params(environment: str | None) -> dict[str, str] | None:
    """Builds the query that scopes a list to one environment, or none for all."""
    return {"environment": environment} if environment else None


def _update_body(value: str | None, description: str | None) -> dict[str, Any]:
    """Builds an update body, rejecting a call that would change nothing."""
    if value is None and description is None:
        raise ValueError("Provide a value or description to update")
    if value is not None:
        _require_value(value)

    body: dict[str, Any] = {}
    if value is not None:
        body["value"] = value
    if description is not None:
        body["description"] = description
    return body


class VariablesService:
    """Synchronous management of a project's variables.

    Reads need the ``variables:read`` scope. Writes need ``variables:write``
    *and* an owner or admin: the API checks the member's role independently of
    the key's scopes, so a plain member's key is refused even when it carries
    the scope.

    Unlike a secret, a variable's value is returned in full on every read. That
    is the difference between the two resources — anything worth hiding belongs
    in :class:`~cosmoner.secrets.SecretsService`.
    """

    def __init__(self, transport: Transport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        project = resolve_project_id(self._config, project_id)
        return f"/v1/projects/{project}/variables"

    def list(
        self, *, environment: str | None = None, project_id: str | None = None
    ) -> dict[str, Any]:
        """Lists the project's variables, values included.

        Args:
            environment: One of ``default``, ``development``, ``staging`` or
                ``production``. Omit to list every environment.
            project_id: Overrides the client-level default project.
        """
        result: dict[str, Any] = self._transport.request(
            "GET",
            self._base_path(project_id),
            params=_environment_params(environment),
        )
        return result

    def get(self, variable_id: str, *, project_id: str | None = None) -> dict[str, Any]:
        """Fetches one variable, value included."""
        _require_variable_id(variable_id)

        result: dict[str, Any] = self._transport.request(
            "GET", f"{self._base_path(project_id)}/{variable_id}"
        )
        return result

    def create(
        self,
        name: str,
        value: str,
        *,
        description: str | None = None,
        environment: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Stores a new variable."""
        _require_name(name)
        _require_value(value)

        body: dict[str, Any] = {"name": name, "value": value}
        if description is not None:
            body["description"] = description
        if environment is not None:
            body["environment"] = environment

        result: dict[str, Any] = self._transport.request(
            "POST", self._base_path(project_id), json=body
        )
        return result

    def update(
        self,
        variable_id: str,
        *,
        value: str | None = None,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Changes a variable's value, its description, or both."""
        _require_variable_id(variable_id)

        result: dict[str, Any] = self._transport.request(
            "PATCH",
            f"{self._base_path(project_id)}/{variable_id}",
            json=_update_body(value, description),
        )
        return result

    def delete(self, variable_id: str, *, project_id: str | None = None) -> None:
        """Permanently removes a variable. The API answers 204, so nothing is returned."""
        _require_variable_id(variable_id)

        self._transport.request("DELETE", f"{self._base_path(project_id)}/{variable_id}")


class AsyncVariablesService:
    """Asynchronous counterpart to :class:`VariablesService`."""

    def __init__(self, transport: AsyncTransport, config: ClientConfig) -> None:
        """Binds the namespace to the client's transport and resolved configuration."""
        self._transport = transport
        self._config = config

    def _base_path(self, project_id: str | None) -> str:
        """Builds the collection route for the resolved project."""
        project = resolve_project_id(self._config, project_id)
        return f"/v1/projects/{project}/variables"

    async def list(
        self, *, environment: str | None = None, project_id: str | None = None
    ) -> dict[str, Any]:
        """Lists the project's variables, values included."""
        result: dict[str, Any] = await self._transport.request(
            "GET",
            self._base_path(project_id),
            params=_environment_params(environment),
        )
        return result

    async def get(
        self, variable_id: str, *, project_id: str | None = None
    ) -> dict[str, Any]:
        """Fetches one variable, value included."""
        _require_variable_id(variable_id)

        result: dict[str, Any] = await self._transport.request(
            "GET", f"{self._base_path(project_id)}/{variable_id}"
        )
        return result

    async def create(
        self,
        name: str,
        value: str,
        *,
        description: str | None = None,
        environment: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Stores a new variable."""
        _require_name(name)
        _require_value(value)

        body: dict[str, Any] = {"name": name, "value": value}
        if description is not None:
            body["description"] = description
        if environment is not None:
            body["environment"] = environment

        result: dict[str, Any] = await self._transport.request(
            "POST", self._base_path(project_id), json=body
        )
        return result

    async def update(
        self,
        variable_id: str,
        *,
        value: str | None = None,
        description: str | None = None,
        project_id: str | None = None,
    ) -> dict[str, Any]:
        """Changes a variable's value, its description, or both."""
        _require_variable_id(variable_id)

        result: dict[str, Any] = await self._transport.request(
            "PATCH",
            f"{self._base_path(project_id)}/{variable_id}",
            json=_update_body(value, description),
        )
        return result

    async def delete(self, variable_id: str, *, project_id: str | None = None) -> None:
        """Permanently removes a variable. The API answers 204, so nothing is returned."""
        _require_variable_id(variable_id)

        await self._transport.request(
            "DELETE", f"{self._base_path(project_id)}/{variable_id}"
        )
